File: easy/Remove_Duplicated_from_Sorted_Array.js
//  Given an integer array nums sorted in non-decreasing order, remove the duplicates in-place such that each unique element appears only once. The relative order of the elements should be kept the same.
//
// Since it is impossible to change the length of the array in some languages, you must instead have the result be placed in the first part of the array nums. More formally, if there are k elements after removing the duplicates, then the first k elements of nums should hold the final result. It does not matter what you leave beyond the first k elements.
//
// Return k after placing the final result in the first k slots of nums.
//
// Do not allocate extra space for another array. You must do this by modifying the input array in-place with O(1) extra memory.

// Solution
//
// Loop throught the array
// Check if each value is equal to the next value in the array
// If it is equal remove it from the array and decrement the index
// If it is not equal move to the next value and repeat
// There is no need to return any value as we are modifying the num array in place
// e.g using a side effect

var removeDuplicates = function(nums) {
    for(let i = 0; i < nums.length; i++){
        if(nums[i] === nums[i+1]){
            nums.splice(i, 1);
            i--;
        }
    }
};

//nums = [0,0,1,1,1,2,2,2,3,3,3,4]
nums = [1,1,2]

removeDuplicates(nums);

console.log("new arr:", nums)

File: easy/Remove_Duplicated_from_Sorted_Array.py
#! /usr/bin/env python3

# This file will need to have the execute permission set
# chmod +x ./LongestCommonPrefix.py


#------------------------------------------------------------------------------#
#                                                                              #
#  Remove duplicates, in place, from an array without using additional memory  #
#                                                                              #
#------------------------------------------------------------------------------#

class Solution(object):
    def removeDuplicates(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        lp = 1

        if len(nums) == 0:
            return 0

        # We use 2 pointers, a right pointer to loop through each integer in the array and a left pointer to pause on each unique value
        # Both pointers initialised to 1
        for rp in range(1, len(nums)):
            # Check each number against the preceding value, if they are not the same we have a new unique value
            if nums[rp] != nums[rp - 1]:
                # Move the new unique number to the position of the left pointer, ie sort it, and then increment the left pointer and continue
                nums[lp] = nums[rp]
                lp += 1
        return lp


s = Solution()
nums = [0,0,1,1,1,2,2,3,3,4]
res = s.removeDuplicates(nums)
print(f'{res} is the number of unique int in {nums}')
